### DISCUSSION TITLE
Globbing for "/**/foo" fails

### DISCUSSION THREAD

**Discussion by tom-dudley** [cake-build/cake#3155] [Q&A]
I'm looking at some of the Globbing code and the following new test fails when I don't think it should.
`VisitRecursiveWildcardSegment` in `GlobVisitor` returns early as the `if` block isn't entered due to `context.Path` being empty. I think that's the cause.

```
[Fact]
public void Should_Parse_Glob_Expressions_With_Root_Then_Directory_Wildcard()
{
    // Given
    var fixture = new GlobberFixture();

    // When
    var result = fixture.Match("/**/*.Test.dll");

    // Then
    Assert.Equal(1, result.Length);
    Assert.ContainsFilePath(result, "/Working/Foo.Bar.Test.dll");
}
```

### What You Are Seeing?
When adding this test, it fails.

### What is Expected?
I think this should pass. (I think it identifies an existing bug)

### What version of Cake are you using?
Latest on `develop` - `55dcb81a9de3a43d36fb0516113aaa175ea58232`

### Are you running on a 32 or 64 bit system?
64 bit

### What environment are you running on?  Windows? Linux? Mac?
Windows
### Are you running on a CI Server?  If so, which one?
Nope

**Comment by tom-dudley** [comment]
It's possible this actually works fine on an actual filesystem, it may be caused by the mock implementation. I've not tested that.

**Comment by augustoproiete** [comment]
@tom-dudley It's been a while, but following up on this one. It's possible there was a bug either with the `GlobberFixture` or the actual globbing at the time you tested.

I tried reproducing it with the current codebase as of `develop` and it works as expected. Below is your code adjusted to work with the latest version of code, and both Windows and UnixLike tests pass.

![image](https://user-images.githubusercontent.com/177608/108140493-8b56e600-7098-11eb-8244-0eda93000a8d.png)

```csharp
public sealed class ReproCakeBuild_Issue_1610
{
    [Fact]
    public void Should_Parse_Glob_Expressions_With_Root_Then_Directory_Wildcard_On_Windows()
    {
        // Given
        var fixture = GlobberFixture.Windows();

        // When
        var result = fixture.Match("/**/*.Test.dll");

        // Then
        Assert.Equal(2, result.Length);
        AssertEx.ContainsFilePath(result, "C:/Working/Project.A.Test.dll");
        AssertEx.ContainsFilePath(result, "C:/Working/Project.B.Test.dll");
    }

    [Fact]
    public void Should_Parse_Glob_Expressions_With_Root_Then_Directory_Wildcard_On_UnixLike()
    {
        // Given
        var fixture = GlobberFixture.UnixLike();

        // When
        var result = fixture.Match("/**/*.Test.dll");

        // Then
        Assert.Equal(2, result.Length);
        AssertEx.ContainsFilePath(result, "/Working/Foo.Bar.Test.dll");
        AssertEx.ContainsFilePath(result, "/Working/Bar.Qux.Test.dll");
    }
}
```
